fabric-input-events-v1/src/main/java/net/fabricmc/fabric/impl/client/input/FabricMouseImpl.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.client.input;
+
+import java.nio.DoubleBuffer;
+
+import org.lwjgl.glfw.GLFW;
+import org.lwjgl.system.MemoryStack;
+
+import net.minecraft.client.MinecraftClient;
+import net.minecraft.client.util.Window;
+
+import net.fabricmc.fabric.api.client.input.v1.FabricKeyboard;
+
+public class FabricMouseImpl {
+	public static FabricMouseImpl INSTANCE = new FabricMouseImpl();
+
+	private double x = 0.0;
+	private double y = 0.0;
+	private int buttons = 0;
+	private int modKeys = 0;
+
+	private FabricMouseImpl() {
+	}
+
+	public double getX() {
+		return this.x;
```

Comment:
Coordinates of the mouse, how to say that...

Why does the update method calls a GLFW method and not `Mouse#getX` which can be get using a `MinecraftClient` instance?

Reply:
See the comment on FabricKeyboardImpl above. Also, I believe Mouse#getX is only updated once per tick, not per event.

Reply:
Except this will break the feature of one of my mod very badly and I won't be able, in any way (or I have to mixin in FAPI entirely to change that), to fix that.

It literally destroys the possibility of having a virtual mouse (very useful for """splitscreen""" and controller support) and will lock everything to the system mouse, which is arguably bad in this case.

Reply:
Interesting concern. I don't know how your mod feature is implemented, but it sounds like something that should be supported by fabric.

Reply:
The way it's done is by ignoring the GLFW callback and calling the mouse position update manually of `Mouse`:
https://github.com/LambdAurora/LambdaControls/blob/1.16/fabric/src/main/java/me/lambdaurora/lambdacontrols/client/controller/InputManager.java#L73

Not calling the GLFW method when using a virtual mouse is obviously done to be able to unlink entirely the physical mouse from the game, which is needed for "splitscreen" with one instance using entirely the controller with virtual mouse.

Reply:
Should be fixed by d7e6b2f

Arguably there should be a way to inject and modify input events as they come in, but that can come later. Once again, something more of a concern for a potential future `Observable<EventType>` implementation than the `Event<ListenerType>` implementation we have right now.